### ISSUE TITLE
Cookie notice: Opt-in

### ISSUE THREAD

**Issue by BlackbitNeueMedien** [pimcore/pimcore#5087]
Currently the cookie notice is just a tooltip which shows a message like "This website uses cookies" with an accept button. 
In our opinion this is not enough after current jurisdiction of the European Court of Justice (see https://openjur.de/u/2180608.html - this is only in german) - especially for cookies which are not necessary for the usage of the website (e.g. tracking cookies).
In https://stackoverflow.com/questions/10306254/is-it-possible-to-block-cookies-from-being-set-using-javascript-or-php is shown how to prevent cookies being set. 
What do you think - should Pimcore include an opt-in solution for cookies and prevent cookies from being set if the visitor has not accepted cookie usage?

Perhaps we could extend the cookie notice like in https://github.com/ConnectHolland/cookie-consent-bundle and track which cookies have been tried to be set. In the backend we could then have a list of cookie names which have tried to be set with above's implementation. In this list the admin (or anyone having the to-be-created "cookie" right) could specify to which category certain cookie names belong. This way the visitor really could decide which cookie categories are acceptable to him.

**Comment by BlackbitNeueMedien** [comment]
https://www.webtoffee.com/how-to-automatically-block-cookies-using-the-gdpr-cookie-consent-plugin/ does something similar for Wordpress

**Comment by Cruiser13** [comment]
As far as I know necessary cookies (login, cart, checkout) are still allowed without opt-in.
If other cookie opt-ins are needed, we're currently using https://github.com/AmauriC/tarteaucitron.js which plays nicely with pimcore.

**Comment by NiklasBr** [comment]
> Pimcore should support website owners with a legal cookie consent solution. What do you think?

100% agreement.

**Comment by msphn** [comment]
@brusch I'd think a statement here would be great :)

**Comment by NiklasBr** [comment]
@brusch ping :)

**Comment by brusch** [comment]
First of all, thanks for your inputs on this topic.
We also see the need of a generic solution built-in in Pimcore, but currently it's probably the most worse time to implement a new solution. 

You may ask why?
Well, I'm not a lawyer but, from what I've read, there's a problem with the recent judgement of the European Court of Justice, because it's just a decision for one particular thing, but it doesn't answer all the other open questions where a jurisdiction is missing. Just think of the question which cookies are indispensable and which ones not and there are a lot more things that need clarification. 
And then there's another problem with the local jurisdiction of the EU member states (remember, the cookie directive it not a law, it needs to be ratified by the local legislation), first and foremost Germany, where there are some more outstanding judgments for important questions from the BGH (Federal Court of Justice). 

So in my opinion we should wait a bit longer until we implement a new solution for this very important topic.

Other opinions?

**Comment by Cruiser13** [comment]
I believe that we'll not get a better clarification until eprivacy is coming - which is delayed already. The specifications may vary but not having a built-in solution until eprivacy arrives will lead to issuers with data privacy on pimcore websites for at least the next six months I guess.

For our agency, this is an issue that needs to be solved in 11-2019 - we have our own solution almost ready and are currently looking into pimcore bundles to distribute it or waiting for a pimcore solution making our solution obsolete. And I guess that others - especially in germany - are in the same situation.

**Comment by NiklasBr** [comment]
> Other opinions?

My personal opinion, is that the current solution is *not* legal and that waiting will cause undue risks due to this non-compliance for all users of Pimcore. Waiting is not optional any longer since the enforcement grace period has ended.

It is very simple in Recital 42, and 43:

> Consent should not be regarded as freely given if the data subject has no genuine or free choice or is unable to refuse or withdraw consent without detriment.

The current solution does not allow users the ability to refuse (or withdraw) consent to non-essential data collection. This is a very clear violation, no further waiting for clarifications should be needed here.

> In order to ensure that consent is freely given, consent should not provide a valid legal ground for the processing of personal data in a specific case where there is a clear imbalance between the data subject and the controller, in particular where the controller is a public authority and it is therefore unlikely that consent was freely given in all the circumstances of that specific situation. Consent is presumed not to be freely given if it does not allow separate consent to be given to different personal data processing operations despite it being appropriate in the individual case, or if the performance of a contract, including the provision of a service, is dependent on the consent despite such consent not being necessary for such performance.

**Comment by dkorbginski** [comment]
IMO we can only reasonably implement a very generic solution where the user can choose to opt-in to the different cookie categories, and leave it to the developer to decide what to do with that information, as in:
```
{% if user opted in to analytics cookies %}
  Add Google Analytics
{% endif %}
```

That being said, I think integrating the ConnectHolland/cookie-consent-bundle (or something in that fashion) would probably be reasonable option.

We will never be able to deliver a solution that can automatically decide which cookies to allow or to block based on user consent. As @brush said, its still highly controversial how to categorize cookies, so we have to leave that decision to the implementation.
Regarding the timing, we could just as well implement a generic solution right now. No court ruling will ever make it clear how to deal with every situation. If someone wants to add Google Analytics under the "strictly necessary" category, we should just give him an option to do so.

**Comment by Cruiser13** [comment]
We just published our solution as pimcore bundle here: https://github.com/pringuin/dataprivacybundle

**Comment by brusch** [comment]
Closed, see #5711

**Comment by brusch** [comment]
@Cruiser13 thanks for sharing your solution!
